feat: add Polishing and Summary Features

Jerry23011 opened the pull request:
closes #588 
<img width="472" alt="截屏2024-07-12 21 50 31" src="https://github.com/user-attachments/assets/44b545e8-c83f-4b2f-a44a-a5d956ea0140">

tisfeng commented on Easydict/Swift/Service/LLMDerivs/SummaryService.swift:
```diff
@@ -0,0 +1,97 @@
+            serviceType: .summary
+        )
+
+        let chatHistory = serviceChatMessage(chatParam)
```
"`SummaryService` and `PolishingService` have too much duplicate code; we can extract the prompt  code, add a new method called `chatHistory` to `BaseOpenAIService`, and they would only need to override this method.

tisfeng replied:
Sorry, I forgot there's already a method `serviceChatMessageModels()`, so `SummaryService` and `PolishingService` just need to override it.

tisfeng replied:
I've refactored the code, now we just need to simply rewrite the `chatMessageDicts()` method.

Jerry23011 replied:
Which part of `chatMessageDicts()` do we need to rewrite and what do we want to achieve?

tisfeng replied:
For custom AI tool service, you only need to rewrite `chatMessageDicts` to return a prompt in the format of `[[String: String]]`. 

In the `serviceChatMessageModels` method of `BaseOpenAIService`, we will automatically convert the dictionary format into the `ChatMessage` format required by `OpenAI`.

Please refer to the code for details.

<img width="1301" alt="image" src="https://github.com/user-attachments/assets/00820879-3ba9-4019-a700-247506388926">

tisfeng commented:
I tried `Summary` and it didn't work very well.

First, it should answer in the user's first preferred language, e.g. if I set Chinese as my first preferred language, it should summarize in Chinese at this point.

Second, the summary is too long, the summary should be concise and clear, the prompt may need to be improved.

This is just an experiment to test the AI tool service,  if `Summary` is not good, it's ok to only add `Polishing`.

<img width="496" alt="56570" src="https://github.com/user-attachments/assets/40b1e76c-729b-47b3-b7b1-15c5b897fc24">

<img width="866" alt="45802" src="https://github.com/user-attachments/assets/c0eb5bce-de17-421d-a8c0-404db056bd71">

Jerry23011 commented:
> First, it should answer in the user's first preferred language

That's very true, I added `Your summary should also be in \(sourceLanguage.queryLanguageName).` in the prompts to explicitly tell the model to output the `queryLanguage`

> Second, the summary is too long, the summary should be concise and clear, the prompt may need to be improved.

I added some examples for Chinese, hope it'll improve on performance for Chinese summary
<img width="472" alt="截屏2024-07-13 18 57 25" src="https://github.com/user-attachments/assets/f9ff019d-6c89-4e58-8cad-c65681c52db7">

tisfeng commented:
> First, it should answer in the user's first preferred language, e.g. if I set Chinese as my first preferred language, it should summarize in Chinese at this point.

No, what I mean is, even if the source language of the text here is English, it should still be summarized in Chinese(first preferred language).

<img width="496" alt="55231" src="https://github.com/user-attachments/assets/03ccb22d-0cb3-4793-98f6-3ca3eb9313c9">

Jerry23011 commented:
> Even if the source language of the text here is English, it should still be summarized in Chinese

Can I add a toggle in settings to let users choose if they want to use the preferred language for the summary? I would expect for the summary to be the same as the source language.

tisfeng commented:
I think it would be better to answer these queries in the user's first preferred language, so that it is easier for users to look up any language they are not familiar with, not just Chinese or English。

For example I could use it to summarize articles in Ukrainian.

The same logic is currently used for `Sentence` and `Dictionary`,  it would be better to keep it consistent.

<img width="496" alt="48039" src="https://github.com/user-attachments/assets/bb84faec-1a74-473e-a908-c29c9de2393d">

tisfeng commented:
For these unconventional AI tool services, it's better to set their usage status to `alwaysOff` by default, and only use them manually when needed 91ca4c07

tisfeng commented:
For AI services, prompts are very important. 

These prompts may need further optimization later to achieve optimal results.

tisfeng reviewed (approved):
Looks good.

Jerry23011 commented:
> For AI services, prompts are very important.
> 
> These prompts may need further optimization later to achieve optimal results.

I’ll look it up later and see how we can improve them

phlpsong commented on Easydict/App/Localizable.xcstrings:
```diff
@@ -3834,6 +3850,16 @@
+        "zh-Hans" : {
+          "stringUnit" : {
+            "state" : "translated",
+            "value" : "总结"
```
Looks like this missing the `en` localization.

<img width="337" alt="Screenshot 2024-07-14 at 11 32 29" src="https://github.com/user-attachments/assets/f0b55b78-88d4-4ff4-85af-368beb1cb05f">

Jerry23011 replied:
My bad, thank you

Jerry23011 replied:
Sorry I didn’t take my laptop with me. Please let me fix it tonight

tisfeng replied:
I have added this en localization.

phlpsong replied:
OK, it's my bad, I thought it has been fixed and not check the commit. 🥲

tisfeng commented:
The icons for the `Polishing` and `Summary` services don't look very good, they seem to be too small and not clear, please check.

<img width="448" alt="87209" src="https://github.com/user-attachments/assets/976d95ec-94e3-44ea-ad3c-ebea1df8ae96">

Jerry23011 commented:
> The icons for the `Polishing` and `Summary` services don't look very good, they seem to be too small and not clear, please check.
> 
> <img alt="87209" width="448" src="https://private-user-images.githubusercontent.com/25194972/348515596-976d95ec-94e3-44ea-ad3c-ebea1df8ae96.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA5MzI2MzgsIm5iZiI6MTcyMDkzMjMzOCwicGF0aCI6Ii8yNTE5NDk3Mi8zNDg1MTU1OTYtOTc2ZDk1ZWMtOTRlMy00NGVhLWFkM2MtZWJlYTFkZjhhZTk2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzE0VDA0NDUzOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTkzMTFmMzk2MzQ3YWJhYTQzMmQ1ZmZhMTYxZTNkOGU4MDEzYjkyNWUzODgxN2NmM2IxZGQxZDk3MDRhMzA1ZDUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.ZyYp8ph3YB7BWR4yXnzNBC39CmzrOHK-TzWYNrEhdBU">

Yes indeed, thanks for pointing out. Let me check it tonight

Jerry23011 commented:
> The icons for the `Polishing` and `Summary` services don't look very good, they seem to be too small and not clear, please check.

Fixed with https://github.com/tisfeng/Easydict/pull/609/commits/72ccb48f45a8ca2006fe6745d0df9b13f3742e3f

tisfeng commented:
ok, polishing icon looks good now, but summary icon still looks unclear, are there any other alternatives?

<img width="497" alt="54043" src="https://github.com/user-attachments/assets/4284d2e8-3cd2-4c27-a06b-bfe9cd67814a">

Jerry23011 commented:
They’re generated by DALL·E, let me try some new ones tomorrow

Jerry23011 commented:
I've added a new icon
<img width="472" alt="截屏2024-07-16 14 51 28" src="https://github.com/user-attachments/assets/200e145f-ffa0-47f3-a3cc-c6e27765b815">

tisfeng commented:
ok, looks good now.
